yapapi/rest/market.py

```diff
@@ -77,6 +77,17 @@ async def details(self, force_refresh=False) -> AgreementDetails:
             self._details = AgreementDetails(_ref=await self._api.get_agreement(self._id))
         return self._details
 
+    @property
+    def cached_details(self) -> AgreementDetails:
```

Comment:
I'd argue this should return `Optional[AgreementDetails]` and a `RuntimeError` should be raised in `WorkContext#_agreement_details` instead. My logic here is the following: it's valid for an `Agreement` object to have no cached details yet, but it's invalid for a `WorkContext` object to not be able to return `provider_name` or `provider_id` (which both rely on agreement details being available).

Reply:
Your argument is valid, but I have two : )

1. As long as `cached_details` is used only in `WorkContext._agreement_details` then there's no difference, but I expect more entities to use `cached_details` in the future (e.g. consumers of the events with attached Agreement). In that case, we'll have to duplicate this RuntimeError in every consumer (or add `assert`s etc).
2. Currently there's just no way of having an agreement without `cached_details` anywhere except deep internals of `AgreementsPool`. This is speaks for hiding this special-strange-case-handling as deeply as possible (so e.g. you don't have to think about it when reading `WorkContext` code).

And also, there's `_details` attribute than can be used if we want really "details as `Agreement` knows them". So for me this is just sort of a convenience helper function used in specific situations where we know it's OK to use it.

In general, from the interface POV every `-> Optional[` is a cost, because we force the interface user to deal with `None`s. If there is more than one user they'll probably be dealing with it somehow similar -> it's nice to have a non-optional accessor with some default `None` handling.

If this doesn't convince you then I will make the change, since - as said before - your argument is valid and I'm 100% sure my arguments are better : )

Reply:
I'm not fully convinced, but I admit it's a minor thing and given the current assumptions in the project (most importantly `AgreementsPool`) I think it's fine this way. :shrug: